lib/Charts/ChartRenderer.js

```diff
@@ -0,0 +1,423 @@
+function findSelectedData(data, x) {
+    // For each chart line (pointArray), find the point with the closest x to the mouse.
+    const closestXPoints = data.map(line => line.points.reduce((previous, current) =>
+        Math.abs(current.x - x) < Math.abs(previous.x - x) ? current : previous
+    ));
+    // Of those, find one with the closest x to the mouse.
+    const closestXPoint = closestXPoints.reduce((previous, current) =>
+        Math.abs(current.x - x) < Math.abs(previous.x - x) ? current : previous
+    );
+    const nearlyEqualX = (thisPoint) => (Math.abs(thisPoint.x - closestXPoint.x) < threshold);
+    // Only select the chart lines (pointArrays) which have their closest x to the mouse = the overall closest.
+    const selectedPoints = closestXPoints.filter(nearlyEqualX);
+
+    const isSelectedArray = closestXPoints.map(nearlyEqualX);
+    const selectedData = data.filter((line, i) => isSelectedArray[i]);
+    selectedData.forEach((line, i) => {line.point = selectedPoints[i];});  // TODO: this adds the property to the original data - bad.
+    return selectedData;
+}
+
+function hilightData(selectedData, scales, g) {
+    const verticalLine = g.select('.selection').selectAll('line').data(selectedData.length > 0 ? [selectedData[0]] : []);
+    verticalLine.enter().append('line')
+        .merge(verticalLine) // New pattern in d3 v4.0 https://github.com/d3/d3/blob/master/CHANGES.md#selections-d3-selection
+        .attr('x1', d => scales.x(d.point.x))
+        .attr('y1', d => scales.y[d.units || Scales.unknownUnits].range()[0])
+        .attr('x2', d => scales.x(d.point.x))
+        .attr('y2', d => scales.y[d.units || Scales.unknownUnits].range()[1]);
+    verticalLine.exit().remove();
+
+    const selection = g.select('.selection').selectAll('circle').data(selectedData);
+    selection.enter().append('circle')
+        .merge(selection)
+        .attr('cx', d => scales.x(d.point.x))
+        .attr('cy', d => scales.y[d.units || Scales.unknownUnits](d.point.y))
+        .style('fill', d => defined(d.color) ? d.color : '');
+    selection.exit().remove();
+}
+
+function highlightDataAndShowTooltip(hasData, data, state, scales, g, tooltip) {
+    if (!hasData) {
+        return;
+    }
+    const localCoords = d3Mouse(g.nodes()[0]);
+    const hoverX = scales.x.invert(localCoords[0]);
+    const selectedData = findSelectedData(data, hoverX);
+    hilightData(selectedData, scales, g);
+    const chartBoundingRect = g['_parents'][0].getBoundingClientRect();  // Strangely, g's own width can sometimes be far too wide.
+    Tooltip.show(Tooltip.html(selectedData), tooltip, state.tooltipSettings, chartBoundingRect);
+}
+
+function unhilightDataAndHideTooltip(g, tooltip) {
+    g.select('.selection').selectAll('circle').data([]).exit().remove();
+    g.select('.selection').selectAll('line').data([]).exit().remove();
+    Tooltip.hide(tooltip);
+}
+
+module.exports = new ChartRenderer();
```

Comment:
Not sure about returning a `new ChartRenderer()` - none of our other classes do that, and it might trip someone up later.